command pg build-settings-type added

yuki-tylar opened the pull request.

pmvald commented on packages/@pinglue/mongodb-pl/src/settings.ts:
```diff
@@ -0,0 +1,20 @@
+/* tslint:disable */
```
this should be moved to the test file. We also need to removed all these comments.  

In general we should not modify any package other than the one we are working on. All these testing should be done with the fake fs as explained in my last post.

pmvald commented on packages/@pinglue/cli/src/actions/build-settings-type.ts:
```diff
@@ -0,0 +1,74 @@
+import path from 'path';
+import yaml from 'yaml';
+import jsonToTypeScript from 'json-schema-to-typescript';
+import { print, style } from "@pinglue/print/node";
```
these two should come from the settings. Please check other action files.

pmvald commented on packages/@pinglue/cli/src/actions/build-settings-type.ts:
```diff
@@ -0,0 +1,74 @@
+    // Find registers.yaml file and read
+    print('Loading pg.yaml file...\n');
+
+    const filepath = path.join('pg.yaml');
```
path.join is needed when you have more than one parts. right now `path.join('abs') === 'abc'`

pmvald commented on packages/@pinglue/cli/src/actions/build-settings-type.ts:
```diff
@@ -0,0 +1,74 @@
+
+export default function (settings: CliActionSettings) {
+  return async (routeName: string, options: Options) => {
+    routeName = routeName || '/';
```
`./`

pmvald commented on packages/@pinglue/cli/src/actions/build-settings-type.ts:
```diff
@@ -0,0 +1,74 @@
+    try {
+      settingsType = await promise;      
+    } catch (error) {
+      print.error('Could not compile settings data into typescript format properly. It may be caused because pg.yaml has incorrect format\n');
```
`print.error(message, data)`  -> data will be automatically printed with `errorObj` style

pmvald commented on packages/@pinglue/cli/src/actions/build-settings-type.ts:
```diff
@@ -0,0 +1,74 @@
+
+    print.success('Compile done. exporting...\n');
+
+    const dirToWrite = path.join('./src');
```
not needed

pmvald reviewed (changes requested):
Now you can add a few tests following `build-channels-types.spec.ts`

yuki-tylar commented:
updated as you mentioned.
I also modified build-settings-types because these two files have some same mistake.

pmvald commented:
Some parts seem to be the old code. Right now your branch is three commits behind the main. So first do a sync with the main. I think some new changes in main must have affected your work as well. As a general rule, always watch for new commits in main and sync to them to avoid conflict and building on old code.

yuki-tylar commented:
I merged `main` into the branch and I tried `5pm run build-ts` in `cli` folder to check if it works properly. but the command `build-ts` got errors such as `Module '"@pinglue/utils/testing/replace-module"' has no exported member 'resetModules'.`

These are related to testing-unit, so I thought I might miss some dependencies. (maybe Jasmine / Mocha or something)

So, I tried these commands at root
```
npm install <-- ok
npm run bt <-- NG (see ERROR1 below)
npm run bt --force <-- ok
npm run build-ts <-- NG (see ERROR2 below)
```

And went back to `cli` folder and did `npm run build-ts`. I still got same error.
Am I still missing something to compile with testing unit properly?


here is ERROR1
```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: pg-monorepo@undefined
npm ERR! Found: typescript@4.5.5
npm ERR! node_modules/typescript
npm ERR!   typescript@"4.5.5" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer typescript@"4.0.x || 4.1.x || 4.2.x || 4.3.x || 4.4.x" from typedoc@0.21.10
npm ERR! node_modules/typedoc
npm ERR!   typedoc@"^0.21.4" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

here is ERROR2
```
src/testing/fake-fs.ts(4,13): error TS2339: Property 'native' does not exist on type '{ (path: PathLike, callback: TCallback<TDataOut>): any; (path: PathLike, options: string | IRealpathOptions, callback: TCallback<...>): any; }'.
@pinglue/utils: src/testing/fake-fs.ts(5,4): error TS2339: Property '__fake' does not exist on type 'IFs'.
```

pmvald commented:
right now this branch has only two commits. Maybe you did the merge on your local computer. So you can push the commit here and we'll look into it. This must be a basic problem for sure.

yuki-tylar commented:
I pushed the new branch in which main is merged.

pmvald commented:
Just by looking at the page I can see that the problem is the merge was done properly. You see it says there is conflict
![Screenshot](https://user-images.githubusercontent.com/31361906/155630421-f6c5023a-2864-4891-85d2-e1874b559d09.png)
:
Now I'll try to find more details. But in general you can refer to the guides I posted here and there. There must be some guide for merging and resolving conflicts. We all need to know how to sync branches with main, this is a common thing we do all the time.

pmvald commented:
Error 2 is easy, we just need to make the extra fields as indices. I just did it now at [this commit](https://github.com/pinglue/pg-repo/commit/911efa0f0c2a1033b683154f97ab180fe4c7442f) - you can do another sync

pmvald commented:
For Error 1, our version of typedoc package is old and it does not work with typescript 4.5.5 - but the new version of typedoc (0.22.12) works with 4.5.x as well, [see here](https://github.com/TypeStrong/typedoc/blob/master/package.json) - so you can star ta new branch and update typedoc version to 0.22.12 for all the packages in pg-repo - using lerna from monorepo root:

```
lerna add typedoc@0.22.12 --scope=pinglue --scope=@pinglue/*
```
Then make a PR!

yuki-tylar commented:
I finally did run properly. Thank you!
I made some small changes at `build-channels.types.ts`.

yuki-tylar commented:
Removed following types files in mongodb-pl
* settings.ts
* channels-types.ts 
* route1/channels-types.ts

Removed unnecessary `promise` check

pmvald reviewed (commented):
test file? (build-settings-type.spec.ts)

yuki-tylar commented:
* Added spec file (I don't know why but I could not create fake `src/settings.ts` file by running the command, so I added the fake empty `src/settings.ts` file manually in fake volume.
* removed comment `generated by Pinglue Cli automatically`. from build-settings-type and build-channels-types

pmvald commented on packages/@pinglue/cli/src/actions/build-settings-type.spec.ts:
```diff
@@ -0,0 +1,161 @@
+    expect(fakeConsole.last().type).to.equal("success");
+
+    const regex = new RegExp(/\s+/g);
+    let returnValue = fakeFs.readFileSync("src/settings.ts", "utf8").toString().trim().replace(regex, ' ');
+    let expectedValue = SETTINGS_TYPE.trim().replace(regex, ' ');
```
Modifying expected and return values in this way is potentially dangerous. You want to see if two codes are equal ignoring indentations? This is a complicated task to do. So lets do it easy way: compare the return value with what it should "exactly" be (including all the white spaces/indentations) 

we also usually hardcode the expected value for each test. So the variable SETTINGS_TYPE is better to define within this test instead of being global (each test will have different values for these variables)
